### ISSUE TITLE
Draggable: auto scroll bug introduced in gsap 1.18.1

### ISSUE THREAD

**Issue by jamesbrobb** [greensock/GreenSock-JS#132]
Bower's just auto updated me to 1.18.2 and i've noticed that there's now an issue with Draggable... which at this point i'm presuming relates to autoScroll. If i have a container with scrollable content and try to drag below the page fold, it begins to move and then jumps the container back to the top. So i'm no longer able to drag content in scrolling containers.

I'm just flagging this with a brief description (as i don't have time to investigate further this evening) incase anything immediately comes to mind?

I'll do some further investigation tomorrow/saturday and put together an example.

**Comment by jamesbrobb** [comment]
I've done some preliminary testing in chrome to see if i can pin point what's causing the issue. My use case is a vertical list of draggable items that's locked to drag on the y axis and with autoScroll.

I started by looking at the changes to Draggable.js, specifically the code block starting at line 1153.

1.18.0

When i drag down vertically, the following `if` runs once the container needs to scroll and this is the only block that runs when dragging that direction.

`if (pointerY > rect.bottom - 40 && (gap = parent._gsMaxScrollY - parent.scrollTop)) {`

Then when dragging up only this block runs

`} else if (pointerY < rect.top + 40 && parent.scrollTop) {`

1.18.1

However in 1.18.1, as soon as i start dragging down (not just when the container needs to scroll), both of the equivalent blocks of code are getting fired one after the other, causing the container to jerk up and down and preventing the autoScroll from functioning correctly.

It appears to relate to the `pointerY` value, as this is continually alternating between being the expected  value and then 0?

**Comment by jamesbrobb** [comment]
The issue is caused by the change in the first line of code in the `render` method of Draggable.js

**1.18.0**

`if (self.autoScroll && self.isDragging && (dirty || checkAutoScrollBounds)) {`

**1.18.1**

`if (self.autoScroll && (checkAutoScrollBounds || (self.isDragging && dirty))) {`

----------------------------------------------------------------------------------------------------

In the original working version `self.isDragging` prevented the function body running, in the same way that it's used in the `updateScroll` method to prevent it from continually running when `self.pointerY` is reset to 0.

So in the version that the bug is introduced, when `self.pointerY` is reset to 0 (although i'm unable to work out where/how this occurs?) and `self.isDragging` is `undefined`, the `render` method is still getting run, causing the incorrect behaviour to occur.

**Comment by jamesbrobb** [comment]
Also i've just noticed that using `https://cdnjs.cloudflare.com/ajax/libs/gsap/latest/utils/Draggable.min.js` does not load the latest version

**Comment by jamesbrobb** [comment]
The issue i'm seeing is caused by the highlighted changes to the Draggable `update` method.
```javascript
this.update = function(applyBounds, ignoreExternalChanges) {
    var x = self.x,
        y = self.y;
    updateMatrix();
    if (applyBounds) {
        self.applyBounds();
    } else {
        if (dirty && ignoreExternalChanges) {
            render();
        }
        syncXY(true);
    }
    if (self.isPressed && ((allowX && Math.abs(x - self.x) > 0.01) || (allowY && (Math.abs(y - self.y) > 0.01 && !rotationMode)))) {
        recordStartPositions();
    }
    // THIS ADDITION IN 1.18.1 CAUSES THE ISSUE
    if (self.autoScroll) {
        _recordMaxScrolls(target.parentNode);
        checkAutoScrollBounds = true;
        render();
    }
    /////////////////////////////////////////////////////////////////////////////////
    return self;
};
```
----------------------------------------------------------------------------------------------------------------

I have an `onDrag` handler that checks the current position of the element being dragged and then repositions the other elements in the list if necessary. To reposition an element i use this logic.

```javascript
TweenLite.to(element, 0.3, {
    css: {
        transform: 'translate3d(0,' + (y - props.y) + 'px,0)'
    },
    onUpdate: draggable.update
});
```

I'd read somewhere (possibly in your docs?) that to keep the position of each `Draggable` object up to date when repositioning, the `onUpdate` handler could be used to call `draggable.update`. Due to the addition in 1.18.1 i've highlighted , this callback now also continually calls the `render` function, causing the behaviour i'm seeing.

**Comment by jamesbrobb** [comment]
Also i've just noticed that using `https://cdnjs.cloudflare.com/ajax/libs/gsap/latest/utils/Draggable.min.js` does not load the latest version, it loads 0.14.1

**Comment by jackdoyle** [comment]
Thanks for the details, @jamesbrobb. Very helpful. A few comments/questions:

1) Are you saying that you try to tween he target's position WHILE it is being dragged? For example, the user could be in the middle of dragging, and you start animating the dragging element to a different position and yet you want the user's mouse/finger movement to ALSO control the element simultaneously? If I understood it correctly, it seems counterintuitive and I wouldn't really expect that to work since two different inputs are fighting with each other, but maybe I misunderstood.

2) Do you have a reduced test case, perhaps in a codepen or jsfiddle? That'd be super helpful for troubleshooting and making sure I'm solving the right problem. 

Regarding the "/latest/" CDN directory, that's unfortunately out of our control and frankly it's pretty frustrating. We've talked to the CDNJS and CloudFlare folks MANY times about this, and the CloudFlare team (who is ultimately responsible) refuses to support the "/latest/" directory now. CDNJS is great and has tried hard to help, but the Cloudflare team won't do the cache purge that would be required to drop a new version in there. The whole industry seems to have moved away from the whole "/latest/" directory thing anyway, so we no longer support it either. We removed mention of that on our site. I really wish we didn't have to stop supporting it because it feels like a broken promise to users, but unfortunately we just don't have any control over that. We have pestered them many, many times. Sorry about the hassle/confusion.

**Comment by jamesbrobb** [comment]
Hi @jackdoyle thanks for the quick response.

1) No not the target, the position of all the other draggable elements in the list (or rather the one that needs to be repositioned) to reposition them when the user drags the target over them. So similar UI functionality to this (but not this code):

http://codepen.io/anon/pen/vLgXrE

So if i don't call `update` on the `Draggable` object associated with the element that gets moved/tweened, then the Draggable object's x and y get's out of sync with the elements x and y. So i'm using it for the purpose described in the docs. I think?

2) i'll try and put something together tomorrow. The code's coupled to Angular at the moment, so i need to remove that dependency to simplify it. Which i was trying to avoid :)

------------------------------------------------------------------------------------------------------------------------

No problem about the cdn, i can imagine how annoying it is. I use bower myself, but just noticed it when i was tweaking the codepen example above and thought i'd mention it.

**Comment by jamesbrobb** [comment]
I'm afraid i haven't had time to decouple my logic from Angular, but i've simplified it as much as i can, so hopefully it's clear what's going on.

The function where the issue is caused is `_positionElements` (line 241). Please let me know if anything needs explaining.

[1.18.0 example](https://codepen.io/anon/pen/NxdaxX)
[1.18.1 example](https://codepen.io/anon/pen/MKJEVR)

**Comment by jackdoyle** [comment]
Thanks so much for the examples. **SUPER** helpful. And sorry about the delayed response. It took a little while to track down a solid solution that also resolves the original issue that the new code was supposed to fix. I think I've got it nailed now. Feel free to take a peek at a preview of the upcoming release (uncompressed): https://s3-us-west-2.amazonaws.com/s.cdpn.io/16327/Draggable-latest-beta.js

It seems to work great in your demo.

**Comment by jamesbrobb** [comment]
Hey @jackdoyle, sorry for the delay in getting back to you.

Great work, that looks like it's fixed it!

[1.18.2 example](https://codepen.io/anon/pen/GoMqQj)

Although bizarrely it's introduced another issue in my actual project that doesn't occur in the example, whereby the further the page scrolls down, the further the element being dragged moves away from the mouse. So the element being dragged moves at a slower speed than the scroll?

I'm wondering if it relates to the layout, as it's more complex than the example and the scrolling container's buried inside several other elements. So feel free to close this and i'll open another issue if i find it relates to GSAP.

Thanks again for your work fixing this, it's much appreciated.

**Comment by jackdoyle** [comment]
Yeah, I'd be very interested to see an example of the new drifting issue you mentioned. Any chance you could craft a reduced test case in codepen or something? I definitely want to fix any bugs - I just need a little help seeing the issue.

**Comment by jamesbrobb** [comment]
yeah no problem, i'll see if i can reproduce it in a codepen tomorrow

**Comment by rmarscher** [comment]
I was going crazy over this one before realizing it was a known issue in `Draggable`. I had fixed it by checking if `self.pointerX` and `self.pointerY` was zero during the `render`... but it looks like the fix @jackdoyle made was to set `checkAutoScrollBounds = self.isDragging;` in `update`. That's cleaner. All works great after the fix. Thanks.

**Comment by jackdoyle** [comment]
@rmarscher sorry about that :) 

So can I go ahead and close this issue now that it has been addressed in 1.18.2? Anyone else seeing a problem here?

**Comment by rmarscher** [comment]
@jackdoyle no problem. Thanks for the awesome product. The fix is in Draggable 0.14.4 and I don't think that has been shipped in a release yet. My copy of 1.18.2 from npm has Draggable 0.14.3.

**Comment by jackdoyle** [comment]
Should be all set in the GSAP 1.18.3 push. Thanks again.
